kymatio/scattering2d/tests/test_torch_scattering2d.py

```diff
@@ -0,0 +1,376 @@
+                    backend.cdgmm(torch.empty(3, 4, 5, 2), torch.empty(4, 5, 1).cuda())
+                assert "cuda" in exc.value.args[0].lower()
+            elif backend.name=='torch':
+                with pytest.raises(TypeError) as exc:
+                    backend.cdgmm(torch.empty(3, 4, 5, 2), torch.empty(4, 5, 1).cuda())
+                assert "both" in exc.value.args[0].lower()
+
+class TestFFT:
+    @pytest.mark.parametrize("backend", backends)
+    def test_fft(self, backend):
+        x = torch.rand(4, 4, 1)
+        with pytest.raises(TypeError) as record:
+            backend.fft(x)
+        assert ('complex' in record.value.args[0])
+
+        x = torch.randn(4, 4, 2)
+        y = x[::2, ::2]
+
+        with pytest.raises(RuntimeError) as record:
+            backend.fft(y)
+        assert ('must be contiguous' in record.value.args[0])
+
+
+class TestScattering2DTorch:
+    def reorder_coefficients_from_interleaved(self, J, L):
+        # helper function to obtain positions of order0, order1, order2 from interleaved
+        order0, order1, order2 = [], [], []
+        n_order0, n_order1, n_order2 = 1, J * L, L ** 2 * J * (J - 1) // 2
+        n = 0
+        order0.append(n)
+        for j1 in range(J):
+            for l1 in range(L):
+                n += 1
+                order1.append(n)
+                for j2 in range(j1 + 1, J):
+                    for l2 in range(L):
+                        n += 1
+                        order2.append(n)
+        assert len(order0) == n_order0
+        assert len(order1) == n_order1
+        assert len(order2) == n_order2
+        return order0, order1, order2
+
+    @pytest.mark.parametrize("backend", backends)
+    @pytest.mark.parametrize("device", devices)
+    def test_Scattering2D(self, backend, device):
+        test_data_dir = os.path.dirname(__file__)
+        data = torch.load(os.path.join(test_data_dir, 'test_data_2d.pt'))
+
+        x = data['x']
+        S = data['Sx']
+        J = data['J']
+
+        # we need to reorder S from interleaved (how it's saved) to o0, o1, o2
+        # (which is how it's now computed)
+
+        o0, o1, o2 = self.reorder_coefficients_from_interleaved(J, L=8)
+        reorder = torch.from_numpy(np.concatenate((o0, o1, o2)))
+        S = S[..., reorder, :, :]
+
+        pre_pad = data['pre_pad']
+
+        M = x.shape[2]
+        N = x.shape[3]
+
+        # Then, let's check when using pure pytorch code
+        scattering = Scattering2D(J, shape=(M, N), pre_pad=pre_pad, backend=backend, frontend='torch')
+        Sg = []
+        x = x.to(device)
+        scattering.to(device)
+        S = S.to(device)
+        if backend.name == 'torch_skcuda' and device == 'cpu':
+            with pytest.raises(TypeError) as ve:
+                Sg = scattering(x)
+            assert "cuda" in ve.value.args[0]
+        else:
+            Sg = scattering(x)
+            assert torch.allclose(Sg, S)
+
+        # check also the default backend
```

Comment:
"Isn't this already done above since we parametrize over backend?" Comment by @janden .

Reply:
it is, but too much test is fine no?

Reply:
Removing.